shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -1924,6 +1929,37 @@ - (void)pressesCancelled:(NSSet<UIPress*>*)presses
   [_textInputPlugin.get().viewController pressesCancelled:presses withEvent:event];
 }
 
+#pragma mark - Undo/Redo support
+
+- (void)ensureUndoEnabled API_AVAILABLE(ios(9.0)) {
+  if (![self.undoManager canUndo]) {
+    self.undoManager.groupsByEvent = NO;
```

Comment:
Is setting this to `NO` and setting it back necessary? Shouldn't `beginUndoGrouping` be enough?

Reply:
I can take another look at this – I tried a lot of different things to get the undo/redo buttons on the keyboard to show up the first time...this is where it ended up, but that doesn't mean it was all necessary.

Taken with your comment on line 1837, I wonder if this might make more sense in the UndoManagerPlugin – we might be able to make two calls across the channel, one to ensure undo is enabled, then set the state...that would keep this from being too tightly coupled to text input.

Reply:
So I think here, it isn't necessary – in the `FlutterUndoManagerPlugin`, though, it does seem to be. I suspect it has something to do with the fact that we're potentially registering multiple undos in the same event loop, and without `groupsByEvent` off, they still get grouped together by iOS.

I fiddled with trying to move the `ensureUndoEnabled` stuff into `FlutterUndoManagerPlugin`, but couldn't get it to actually work correctly. I'm not sure why, but I think doing it within the same event loop as a key press might be the secret sauce there. The good news is, I think the only thing that's needed for is to get iOS to correctly update the UI of the iPad keyboard on the first keypress. After that, things work as expected. For non-text input use cases, the keyboard wouldn't be up anyway, so that should be less of a concern.

Reply:
Ah interesting thanks for looking into this! Since we're using a dedicated `NSUndoManager` in the plugin, would it work if we just turn `groupsByEvent` off at the beginning once and for all since we'll be manually grouping the events then?

Reply:
So we aren't actually using a dedicated `NSUndoManager` (except in tests), we're just grabbing the one from the `FlutterViewController`.

I do wonder, though, if some of the weirdness that requires the `ensureUndoEnabled` has to do with which `NSUndoManager` is active at the time – I did a bit of reading about it, and if I understand correctly, each `NSResponder` can have it's own `UndoManager`, and iOS walks up the responder chain to get the "active" undoManager.

I'm wondering if we _should_ have our own `FlutterUndoManager` or override the `FlutterTextInputView` to return the one from the `FlutterViewController` or something like that. I ran out of time today before I could play around with that idea, but I'll try to get to it soon.

Reply:
All right, so I think the issue with the `ensureUndoEnabled` bit is that iPadOS only checks the undo manager state after a keypress, so the `ensureUndoEnabled` call allows NSUndoManager to return true from `- canUndo` after the keypress, so iPadOS picks that up and enables the undo button. However, since the calls in `FlutterUndoManagerPlugin` are happening after that, the NSUndoManager state doesn't get checked again.

I was able to find a different workaround by setting the `leadingBarButtonGroups` property on the `UITextInputAssistantItem` from the `FlutterTextView`. That causes iPadOS to check the NSUndoManager state again, so it correctly displays.

Let me know what you think of that approach. Now the FlutterTextInputPlugin is completely uninvolved, which I like.

Reply:
Looked into this, it looks like `-[TUISystemInputAssistantView setNeedsValidation]` triggers the keyboard update. Calling `[self.inputDelegate didChangeSelection: self]` should trigger that.

Reply:
(if you have access to the input view that is).

Reply:
Is the `beginUndoGrouping` pair still needed? It sounded like it was a workaround to trigger a keyboard update?

Reply:
I think the grouping is still needed – the workaround for keyboard state was in the TextInputPlugin where I was registering and immediately removing an undo group. That workaround is replaced by the UITextInputAssistantItem business in FlutterUndoManagerPlugin. Calling selectionDidChange on the inputDelegate seems to be an adequate replacement, though, and seems at least a little more correct than removing and re-adding the input bar items.